_posts/2019-10-23-kafka-authentication-using-oauth-2.0.md

```diff
@@ -0,0 +1,538 @@
+ subjectAltName = DNS:keycloak.keycloak, DNS:keycloak.example.com
+EOF
+
+# create certificate-signing request
+openssl req -new -sha256 \
+  -key keycloak.key \
+  -subj "/CN=keycloak" \
+  -reqexts SAN \
+  -config ssl.cnf \
+  -out keycloak.csr
+
+# Generate the final keycloak certificate, signed by CA
+openssl x509 -req -extfile ssl.cnf -extensions SAN -in keycloak.csr -CA ca.crt -CAkey ca.key -CAcreateserial -out keycloak.crt -days 365 -sha256
+
+```
+
+
+Now, let's store the keycloak certificate files into Kubernetes as a secret.
+
+First, create a new namespace for Keycloak:
+
+    kubectl create ns keycloak
+    
+Then, add the secret:
+
+    kubectl create secret tls tls-keys --cert=./keycloak.crt --key=./keycloak.key -n keycloak
+
+And let's deploy the Keycloak server using the Keycloak project's docker image from Docker Hub:
+
+```
+cat > keycloak.yaml << EOF
+apiVersion: v1
+kind: Service
+metadata:
+  name: keycloak
+  labels:
+    app: keycloak
+spec:
+  ports:
+  - name: http
+    port: 8080
+    targetPort: 8080
+  - name: https
+    port: 8443
+    targetPort: 8443
+  selector:
+    app: keycloak
+  type: LoadBalancer
+---
+apiVersion: apps/v1
+kind: Deployment
+metadata:
+  name: keycloak
+spec:
+  replicas: 1
+  selector:
+    matchLabels:
+      app: keycloak
+  template:
+    metadata:
+      labels:
+        app: keycloak
+    spec:
+      containers:
+      - name: keycloak
+        image: jboss/keycloak
+        env:
+        - name: KEYCLOAK_USER
+          value: "admin"
+        - name: KEYCLOAK_PASSWORD
+          value: "admin"
+        - name: PROXY_ADDRESS_FORWARDING
+          value: "true"
+        ports:
+        - name: http
+          containerPort: 8080
+        - name: https
+          containerPort: 8443
+        readinessProbe:
+          httpGet:
+            path: /auth/realms/master
+            port: 8080
+        volumeMounts:
+        - name: tls-keys
+          mountPath: "/etc/x509/https"
+          readOnly: true
+      volumes:
+        - name: tls-keys
+          secret:
+            secretName: tls-keys
+EOF
+```
+
+    kubectl apply -f keycloak.yaml -n keycloak
+
+You can observe keycloak starting up by attaching to the log:
+
+    POD=`kubectl get pod -n keycloak | grep keycloak | awk '{ print $1 }'`
+    kubectl logs $POD -n keycloak -f
+
+Once Keycloak is up and running, take note of the external IP it is listening on:
+
+    kubectl get services -n keycloak
+    
+You can extract just the IP by:
+
+    KEYCLOAK=`kubectl get svc/keycloak -n keycloak | tail -n 1 | awk '{ print $4 }'`
+
+Now, open it in your browser (you'll need to add a certificate exception because your browser won't trust the CA certificate we created earlier), and login using `admin:admin`:
+
+    open https://$KEYCLOAK:8443/auth
+
+
+In OAuth 2.0 all applications communicating with the authorization server are known as 'clients', and act either as application clients, or as application servers (resource servers).
```

Comment:
Thsi read a bit weird ... all applications are clients, but some clients are servers. But I guess this is correct in OAuth, right?

Reply:
What about:
'When your application servers and clients talk to the OAuth 2.0 authorization server, they are all considered 'clients'. In OAuth 2 terminology your application servers are called 'resource servers' - they are the ones requiring authentication.'

Reply:
That sounds good.